syntax = "proto3";

package northstar_cloud;

// The `NorthStar` provides APIs.

service NorthStarService {
    // User activity
    rpc GetUserRescuePoints(GetUserRescuePointsRequest) returns (GetUserRescuePointsReply);
    rpc AddUser(AddUserRequest) returns (AddUserReply);
    rpc GetUser(GetUserRequest) returns (GetUserReply);
    rpc UpdateUser(AddUserRequest) returns (AddUserReply);
    rpc DeleteUser(AddUserRequest) returns (AddUserReply);
    rpc UploadImage(UploadImageRequest) returns (UploadImageReply);
    rpc GetImage(GetImageRequest) returns (GetImageReply);
}

enum FileType {
    UNKNOWN = 0;
    JPEG = 1;
    PNG = 2;
}

message UploadImageRequest {
    string image_id = 1;
    string image_name = 2;
    FileType image_format = 3;
    bytes image = 4;
    User user = 5;
}

message UploadImageReply{
    bool success = 1;
}

message LatLng {
  // The latitude in degrees. It must be in the range [-90.0, +90.0].
  double latitude = 1;

  // The longitude in degrees. It must be in the range [-180.0, +180.0].
  double longitude = 2;
}

message HealthInfo {
    bool need_medical_support = 1;
}

message User{
    string user_id = 1;
    string user_name = 2;
    string first_name = 3;
    string last_name = 4;
    string phone_number = 5;
    string home_address = 6;
    string email_address = 7;
    string office_address = 8;
    string app_id = 9;
    string app_type = 10;
    string last_updated = 11;
    string created_at = 12;
    LatLng current_location = 13;
    HealthInfo health_info = 14;
    repeated LatLng rescue_locations = 15;
}

message AddUserRequest {
    User user = 1;
}

message AddUserReply {
    bool success = 1;
}

message GetUserRequest {
    string user_id = 1;
    string user_name = 2;
    string app_id = 3;
}

message GetUserReply {
    User user = 1;
}

message GetUserRescuePointsRequest {
    User user = 1;
}

message GetUserRescuePointsReply {
    repeated LatLng rescue_locations = 1;
}

message GetImageRequest {
    string image_id = 1;
    string image_name = 2;
}

message GetImageReply {
    string image_id = 1;
    string image_name = 2;
    FileType image_format = 3;
    bytes image = 4;
    User user = 5;
}